### PR TITLE
Goal priority rework

### PR DESCRIPTION
**Pull request by TarikViehmann**
This draft PR is WIP and not merge-ready as the desired strategy is not finalized yet.

The rough idea is to prioritize working towards started products and to adjust priorities for tasks accordingly.

### CODE REVIEW

**Review by snoato — CHANGES_REQUESTED** [review]
Good changes, some open questions

File: src/clips-specs/rcll-central/goal-production.clp [hunk 1]
```diff
@@ -24,14 +24,66 @@
...
+  (if (eq ?com C1) then
+    (bind ?priority ?*PRODUCTION-C1-PRIORITY*)
+  )
+  (if (eq ?com C2) then
+    (bind ?priority ?*PRODUCTION-C2-PRIORITY*)
+  )
+  (if (eq ?com C3) then
+    (bind ?priority ?*PRODUCTION-C3-PRIORITY*)
+  )
+  (bind ?priority (- ?priority ?*PRODUCTION-PRIO-BASE-STEP*))
```
[the remark this answers is not in the document]
Comment (TarikViehmann) [reply src/clips-specs/rcll-central/goal-production.clp:74]: yes, these are dynamic prio which are only used for support goals. These start at one base prio less (for the respective order) and have some inherent priority to them. That way support goals are always slightly less urgent than the actual production goals.

File: src/clips-specs/rcll-central/production-strategy.clp [hunk 2]
```diff
@@ -1299,7 +1300,55 @@
   (retract ?timer)
 )
 
+; re-evaluate criteria for orders that did not progress yet
+(defrule production-strategy-clear-started-orders
+  "Trigger a re-evaluation of orders to pursue every minute"
+  (declare (salience ?*SALIENCE-LOW*))
+  (time $?now)
+  ?reset-fact <- (wm-fact (key order selection reset) (value FALSE))
+  ?timer <- (timer (name order-selection-reset-timer) (time $?t&:(timeout ?now ?t ?*ORDER-SELECTION-RESET-TIMEOUT*)) (seq ?seq))
```
[the remark this answers is not in the document]
Comment (TarikViehmann) [reply src/clips-specs/rcll-central/production-strategy.clp:1309]: in domain.clp

File: src/clips-specs/rcll-central/production-strategy.clp [hunk 3]
```diff
@@ -1372,26 +1417,177 @@
...
+  (domain-fact (name rs-ring-spec) (param-values ?mps ?ring-color ?ring-spec))
+  (domain-fact (name rs-filled-with) (param-values ?mps ?filled-with))
+  (test (< (sym-to-int ?filled-with) (sym-to-int ?ring-spec)))
+
+  (not (wm-fact (key strategy meta priority increase pay-for-wp args? wp ? mps ?mps $) (value ?old-prio&:(> ?old-prio (dynamic-prio-from-complexity ?com ?ring)))))
+  =>
+  (bind ?priority (dynamic-prio-from-complexity ?com ?ring))
+  ; adjust existing goals that already received a priority update
+  (bind ?old-prio-val 0)
+  (do-for-fact ((?old-prio wm-fact)) (and (wm-key-prefix ?old-prio:key (create$ strategy meta priority increase pay-for-wp))
```
Comment (snoato) [inline src/clips-specs/rcll-central/production-strategy.clp:1539]: little bit confusing as you are overwriting variables here

**Review by snoato — CHANGES_REQUESTED** [review]
one little comment

File: src/clips-specs/rcll-central/production-strategy.clp (hunk 3 above)
Comment (TarikViehmann) [reply src/clips-specs/rcll-central/production-strategy.clp:1539]: the variable is out of scope at this point in time, but you have a point. To increase readability I renamed the one in precondition to higher-prio

**Comment by TarikViehmann** [comment]
Thanks for the review!

**Pull request merged**